ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -863,6 +883,76 @@ public void handleReplicateBlobRequest(NetworkRequest request) throws IOExceptio
             metrics.replicateBlobTotalTimeInMs, null, null, totalTimeSpent));
   }
 
+  /**
+   * Get the Tombstone index entry from the remote replica
+   * @param blobId the blob id which we want to get the tombstone record from
+   * @param remoteHostName the host name of the remote replica
+   * @param remoteHostPort the port of the remote replica
+   * @return a pair of the {@link ServerErrorCode} and the tombstone {@link MessageInfo}.
+   */
+  private Pair<ServerErrorCode, MessageInfo> getTombStoneFromRemoteReplica(BlobId blobId, String remoteHostName,
+      int remoteHostPort) {
+    DataNodeId remoteDataNode = clusterMap.getDataNodeId(remoteHostName, remoteHostPort);
+    if (remoteDataNode == null) {
+      logger.error("ReplicateBlobRequest {} couldn't find the remote host {} {} in the clustermap.", blobId,
+          remoteHostName, remoteHostPort);
+      return new Pair(ServerErrorCode.Replica_Unavailable, null);
+    }
+
+    // ON_DEMAND_REPLICATION_TODO: Add configuration as replicationConfig.replicationConnectionPoolCheckoutTimeoutMs
+    final int connectionPoolCheckoutTimeoutMs = 1000;
+    final String clientId = ON_DEMAND_REPLICATION_CLIENTID_PREFIX + currentNode.getHostname();
+
+    DataInputStream stream = null;
+    try {
+      // Get the index entries from the remote replica with AdminRequest
+      int correlationId = 1;
+      AdminRequest adminRequest = new AdminRequest(AdminRequestOrResponseType.BlobIndex, null, correlationId, clientId);
+      BlobIndexAdminRequest blobIndexRequest = new BlobIndexAdminRequest(blobId, adminRequest);
+
+      ConnectedChannel connectedChannel =
+          connectionPool.checkOutConnection(remoteHostName, remoteDataNode.getPortToConnectTo(),
+              connectionPoolCheckoutTimeoutMs);
+      ChannelOutput channelOutput = connectedChannel.sendAndReceive(blobIndexRequest);
+      stream = channelOutput.getInputStream();
+      AdminResponseWithContent adminResponse = AdminResponseWithContent.readFrom(stream);
+      if (adminResponse.getError() != ServerErrorCode.No_Error) {
+        logger.error("ReplicateBlobRequest failed to get tombstone {} from the remote node {} {} {}", blobId,
+            remoteHostName, remoteHostPort, adminResponse.getError());
+        return new Pair(adminResponse.getError(), null);
+      }
+
+      byte[] jsonBytes = adminResponse.getContent();
+      ObjectMapper objectMapper = new ObjectMapper();
+      StoreKeyJacksonConfig.setupObjectMapper(objectMapper, new BlobIdFactory(clusterMap));
+      Map<String, MessageInfo> messages =
+          objectMapper.readValue(jsonBytes, new TypeReference<Map<String, MessageInfo>>() {
+          });
+      if (messages == null || messages.size() != 1) {
+        logger.error("ReplicateBlobRequest adminRequest for {} from the remote node {} {} return {} entries", blobId,
```

Comment:
can you also print out the content of the map so we know what went wrong.

Reply:
Added the dump.